ibis/expr/timecontext.py

```diff
@@ -0,0 +1,190 @@
+        If neither of the two contexts is a superset of each other, and they
+        share some time range in common, we called them OVERLAP.
+        And NONOVERLAP means the two contexts doesn't overlap at all, which
+        means end1 is earlier than begin2 or end2 is earlier than begin1
+    """
+
+    SUBSET = 0
+    SUPERSET = 1
+    OVERLAP = 2
+    NONOVERLAP = 3
+
+
+def compare_timecontext(left_context: TimeContext, right_context: TimeContext):
+    """Compare two timecontext and return the relationship between two time
+    context (SUBSET, SUPERSET, OVERLAP, NONOVERLAP).
+
+    Parameters
+    ----------
+    left_context: TimeContext
+    right_context: TimeContext
+
+    Returns
+    -------
+    result : TimeContextRelation
+    """
+    left_begin, left_end = left_context
+    right_begin, right_end = right_context
+    if right_begin <= left_begin and right_end >= left_end:
+        return TimeContextRelation.SUBSET
+    elif right_begin >= left_begin and right_end <= left_end:
+        return TimeContextRelation.SUPERSET
+    elif right_end < left_begin or left_end < right_begin:
+        return TimeContextRelation.NONOVERLAP
+    else:
+        return TimeContextRelation.OVERLAP
+
+
+def canonicalize_context(
+    timecontext: Optional[TimeContext],
+) -> Optional[TimeContext]:
+    """Convert a timecontext to canonical one with type pandas.Timestamp
+       for its begin and end time. Raise Exception for illegal inputs
+    """
+    SUPPORTS_TIMESTAMP_TYPE = pd.Timestamp
+    try:
+        begin, end = timecontext
+    except (ValueError, TypeError):
+        raise com.IbisError(
+            f'Timecontext {timecontext} should specify (begin, end)'
+        )
+
+    if not isinstance(begin, SUPPORTS_TIMESTAMP_TYPE):
+        raise com.IbisError(
+            f'begin time value {begin} of type {type(begin)} is not'
+            ' of type pd.Timestamp'
+        )
+    if not isinstance(end, SUPPORTS_TIMESTAMP_TYPE):
+        raise com.IbisError(
+            f'end time value {end} of type {type(begin)} is not'
+            ' of type pd.Timestamp'
+        )
+    if begin > end:
+        raise com.IbisError(
+            f'begin time {begin} must be before or equal' f' to end time {end}'
+        )
+    return begin, end
+
+
+""" Time context adjustment algorithm
+    In an Ibis tree, time context is local for each node, and they should be
+    adjusted accordingly for some specific nodes. Those operations may
+    require extra data outside of the global time context that user defines.
+    For example, in asof_join, we need to look back extra `tolerance` daays
+    for the right table to get the data for joining. Similarly for window
+    operation with preceeding and following.
+    Algorithm to calculate context adjustment are defined in this module
+    and could be used by multiple backends.
+"""
+
+
+@functools.singledispatch
+def adjust_context(op, *clients, timecontext: TimeContext) -> TimeContext:
```

Comment:
Why do we need clients here?

Reply:
Yes, I have explained in responding to your comments to add dispatch adjust_context. It's suggested to add `op` in parameter, and we want to generalize this function so it could be used by multiple backends. In many cases, we will need to call `execute` in `adjust_context` to get time interval (preceding, tolerance etc.) And we will need clients to `execute`.  If we don't add clients here we may want to pass in those executed results of the time interval, just like my previous commit, but that would be harder to add a union dispatch for them. Does it make sense?

Reply:
kk. Find for now.